### PR TITLE
Fix adjoint fluid analyzer

### PR DESCRIPTION
**Pull request by sunethwarna**
This PR introduces following changes.

1. Move adjoint_vms_monolithic_solver and trilinos_adjoint_vmsmonoithic_solver to FluidDynamicsApplication
2. Adapt those two solvers based on the updated PythonSolver
3. Adapt adjoint_fluid_analysis based on the changes in PythonSolver and Analysis_Stage
4. Move unit tests in AdjointFluidApplication to FluidDynamicsApplication

All the test cases in FluidDynamicsApplication including adjoint unit tests are working.

### CODE REVIEW

File: applications/FluidDynamicsApplication/tests/adjoint_cpp_unit_tests.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from KratosMultiphysics import *
+
+class AdjointCPPUnitTests(KratosUnittest.TestCase):
```
Comment (msandre) [inline applications/FluidDynamicsApplication/tests/adjoint_cpp_unit_tests.py]: I get an error that KratosUnittest is not recognized when I run this. But, rather than fixing that, I suggest re-writing this file to be like https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/tests/run_cpp_unit_tests.py.

File: applications/FluidDynamicsApplication/tests/adjoint_vms_element_2d.py [hunk 2]
```diff
@@ -253,4 +263,10 @@ def test_CalculateSensitivityMatrix(self):
         self._assert_matrix_almost_equal(FDShapeDerivativeMatrix, ShapeDerivativeMatrix)
 
 if __name__ == '__main__':
-    KratosUnittest.main()
+    test = AdjointVMSElement2D()
```
Comment (msandre) [inline applications/FluidDynamicsApplication/tests/adjoint_vms_element_2d.py]: Why did you switch from `KratosUnittest.main()`? I liked the old style printout more because it tells you information about the tests that run. This information is missing now.

File: applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        os.chdir(self.scope)
+
+    def __exit__(self, type, value, traceback):
+        os.chdir(self.currentPath)
+
+@KratosUnittest.skipUnless(have_required_applications," ".join(missing_applications_message))
+class AdjointVMSSensitivity2D(KratosUnittest.TestCase):
+
+    def setUp(self):
+        pass
```
Comment (msandre) [inline applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py:41]: There is a lot of print out in these tests coming from the `_ImportModelPart()` of the `python_solver.py`. What about modifying the severity level in the `setUp` and `tearDown` like the following?

```
def setUp(self):
        self._logger_severity = KratosMultiphysics.Logger.GetDefaultOutput().GetSeverity()
        KratosMultiphysics.Logger.GetDefaultOutput().SetSeverity(KratosMultiphysics.Logger.Severity.WARNING)

def tearDown(self):
        KratosMultiphysics.Logger.GetDefaultOutput().SetSeverity(self._logger_severity)
```

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py [hunk 4]
```diff
@@ -74,6 +77,14 @@ def AssambleTestSuites():
     nightSuite.addTest(TimeIntegratedFluidElementTest('testSymbolic'))
     nightSuite.addTest(FluidAnalysisTest('testFluidDynamicsAnalysis'))
     nightSuite.addTest(AdjointFluidTest('testCylinder'))
+    nightSuite.addTest(AdjointVMSElement2D('testCalculateSecondDerivativesLHS'))
+    nightSuite.addTest(AdjointVMSElement2D('testCalculateFirstDerivativesLHS1'))
+    nightSuite.addTest(AdjointVMSElement2D('testCalculateFirstDerivativesLHS2'))
+    nightSuite.addTest(AdjointVMSElement2D('testCalculateSensitivityMatrix'))
+    nightSuite.addTest(AdjointVMSSensitivity2D('testCylinder'))
+    nightSuite.addTest(AdjointVMSSensitivity2D('testOneElement'))
+    nightSuite.addTest(AdjointCPPUnitTests('testSensitivityCPP'))
```
Comment (msandre) [inline applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py]: Here the cpp unit tests can be called like in https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py#L369.

**Review by msandre — CHANGES_REQUESTED** [review]

File: applications/FluidDynamicsApplication/tests/adjoint_cpp_unit_tests.py (hunk 1 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/adjoint_cpp_unit_tests.py]: done :)

File: applications/FluidDynamicsApplication/tests/adjoint_vms_element_2d.py (hunk 2 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/adjoint_vms_element_2d.py]: I changed this in few tests, but some of the tests in FluidDynamicsApplication has some specific method calls. Therefore those are not changed.

File: applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py (hunk 3 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py:41]: This will not filter out the std::cout prints which is still a lot of outputs. I'm not sure how to get rid of these printouts except for using -v2 as an argument when testing.

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py (hunk 4 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py]: Done :)

File: applications/FluidDynamicsApplication/tests/hdf5_io_test.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    def tearDown(self):
+        with ControlledExecutionScope(self.work_folder):
+            # remove hdf5 file
+            self._remove_h5_files("MainModelPart")
+            # remove other generated files
+            kratos_utilities.DeleteFileIfExisting("./square5.time")
+            kratos_utilities.DeleteFileIfExisting("./reference_results.json")
+
+if __name__ == '__main__':
+    test = HDF5IOTest()
```
Comment (msandre) [inline applications/FluidDynamicsApplication/tests/hdf5_io_test.py]: We should use a consistent way to run the tests. Can we call KratosUnittest.main() here too?

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication_mpi.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+# import Kratos
+import KratosMultiphysics
+import KratosMultiphysics.StructuralMechanicsApplication as StructuralMechanicsApplication
```
Comment (msandre) [inline applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication_mpi.py]: Remove StructuralMechanicsApplication.

File: applications/FluidDynamicsApplication/tests/Cavity/input_test_parameters.json [hunk 7]
```diff
@@ -0,0 +1,81 @@
+{
+    "test_settings" : {
+        "execute_solve" : false
```
Comment (msandre) [inline applications/FluidDynamicsApplication/tests/Cavity/input_test_parameters.json]: I think we can remove these `"execute_solve"` settings. They appear to have been replaced by `fluid_analysis_without_solution.py` which is called directly in `hdf5_io_test.py`.

File: applications/FluidDynamicsApplication/tests/Cavity/output_test_parameters.json [hunk 8]
```diff
@@ -0,0 +1,148 @@
+{
+    "test_settings" : {
+        "execute_solve" : true
```
Comment (msandre) [inline applications/FluidDynamicsApplication/tests/Cavity/output_test_parameters.json]: `"execute_solve"`

**Review by msandre — CHANGES_REQUESTED** [review]

File: applications/FluidDynamicsApplication/tests/Cavity/output_test_parameters.json (hunk 8 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/Cavity/output_test_parameters.json]: done :)

File: applications/FluidDynamicsApplication/tests/Cavity/input_test_parameters.json (hunk 7 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/Cavity/input_test_parameters.json]: done :)

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication_mpi.py (hunk 6 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication_mpi.py]: done :)

File: applications/FluidDynamicsApplication/tests/hdf5_io_test.py (hunk 5 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/hdf5_io_test.py]: done :)

File: applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py (hunk 3 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py:41]: Since, this PR is massive, Will it be ok if I add these changes in a separate PR?

**Review by msandre — DISMISSED** [review]

**Comment by msandre** [comment]
Thanks!

File: applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py (hunk 3 above)
Comment (msandre) [reply applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py:41]: ok

**Comment by philbucher** [comment]
@sunethwarna I would also have a look at this PR, can you please wait with merging?

I will try to to it this week

**Comment by sunethwarna** [comment]
@philbucher I will wait

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication_mpi.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
+# import Kratos
+import KratosMultiphysics
+
+# Import Kratos "wrapper" for unittests
+import KratosMultiphysics.KratosUnittest as KratosUnittest
+
+try:
+    import KratosMultiphysics.mpi as KratosMPI
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication_mpi.py:8]: I guess it would be a good idea to also import metris and trilinos here (I should also do this in StructuralMechanics)

The solvers need it anyway and the error will be more clear if it fails here

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py [hunk 10]
```diff
@@ -19,6 +21,10 @@
 from volume_source_test import VolumeSourceTest
 from fluid_analysis_test import FluidAnalysisTest
 from adjoint_fluid_test import AdjointFluidTest
+from adjoint_vms_element_2d import AdjointVMSElement2D
+from adjoint_vms_sensitivity_2d import AdjointVMSSensitivity2D
+from hdf5_io_test import HDF5IOTest
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py:26]: I would use a better name here, this made me think its unrelated to fluid

File: applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py (hunk 3 above)
Comment (philbucher) [reply applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py:41]: Which couts are left?
The tests in StructuralMechanics dont print anything any more if the lvl of severity is increased

File: applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+    def _remove_file(self, file_path):
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/tests/adjoint_vms_sensitivity_2d.py]: It just a test but I would usually go for the naming suggestions of the [Style Guide](https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide#classes-1)

File: applications/FluidDynamicsApplication/tests/adjoint_vms_element_2d.py [hunk 12]
```diff
@@ -146,7 +156,7 @@ def test_CalculateFirstDerivativesLHS1(self):
         self._assign_solution_step_data1(0)
         self._assign_solution_step_data2(1)
 
-    def test_CalculateFirstDerivativesLHS2(self):
+    def testCalculateFirstDerivativesLHS2(self):
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/tests/adjoint_vms_element_2d.py]: I personally would prefer the other name since its more readable

Just my opinion though

File: applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_solver.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+        dt = self._ComputeDeltaTime()
+        new_time = current_time + dt
+
+        self.main_model_part.CloneTimeStep(new_time)
+        self.main_model_part.ProcessInfo[KratosMultiphysics.STEP] += 1
+
+        return new_time
+
+    def InitializeSolutionStep(self):
+        self.solver.InitializeSolutionStep()
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_solver.py:99]: I would highly recommend changing this name!

self.solver is a solving strategy, it becomes really confusing since the class itself is also a solver

=> you can take a look at the solvers in StructuralMechanics

File: applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_solver.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
+from __future__ import print_function, absolute_import, division  # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_solver.py:1]: It seems like this file copies many functionalities from the fluids solver

Maybe it would make sense to derive from it? @jcotela

File: applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_analysis.py [hunk 15]
```diff
@@ -16,179 +16,117 @@ class AdjointFluidAnalysis(AnalysisStage):
     '''Main script for adjoint sensitivity optimization in fluid dynamics simulations.'''
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_analysis.py:16]: this file is basically a copy of the FluidDynAnalysis

I think it could make sense to derive from it and only override the necessary fcts (from what I can see only the CreateSolver fct)

This is how we do it with StructuralMechanics and ContactStructuralMechanics

File: applications/FluidDynamicsApplication/python_scripts/adjoint_vmsmonolithic_solver.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
+from __future__ import print_function, absolute_import, division  # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/adjoint_vmsmonolithic_solver.py:1]: same here again

how abt multiple inheritance form `AdjointFluidSolver` AND `VMSMonolithicFluidSolver`?

This would simplify this file a lot which imo is a valid reason for multiple inheritance

File: applications/FluidDynamicsApplication/python_scripts/compute_drag_process.py [hunk 17]
```diff
@@ -2,17 +2,18 @@
 import KratosMultiphysics.FluidDynamicsApplication as KratosFluid
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/compute_drag_process.py:2]: Changes to this file are not related to this PR and should usually be done in a separate PR

File: applications/FluidDynamicsApplication/python_scripts/steady_navier_stokes_solver_vmsmonolithic.py [hunk 18]
```diff
@@ -83,7 +83,9 @@ def Initialize(self):
         (self.solver).SetEchoLevel(self.settings["echo_level"].GetInt())
         (self.solver).Check()
 
-        self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.DYNAMIC_TAU, self.settings["dynamic_tau"].GetDouble())
-        self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.OSS_SWITCH, self.settings["oss_switch"].GetInt())
+        if self.settings["stabilization"].Has("dynamic_tau"):
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/steady_navier_stokes_solver_vmsmonolithic.py:86]: Same comment as for the drag process, should be done in a separate PR

File: applications/FluidDynamicsApplication/python_scripts/trilinos_adjoint_vmsmonolithic_solver.py [hunk 19]
```diff
@@ -2,24 +2,27 @@
 ## Importing the Kratos Library
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/trilinos_adjoint_vmsmonolithic_solver.py:2]: same comment as for the serial version, lots of copy pasted code

**Review by philbucher — COMMENTED** [review]
Overall nice work, have a look at my comments

I have a major concern though regarding the copy pasting of code form the fluid to the adjoint-fluid solver/stage

I think it makes sense to use inheritance as suggested in my comments
Guess I don't have to list the disadvantages of copy pasting code :)

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py (hunk 10 above)
Comment (jcotela) [reply applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py:26]: Well, we are actually testing the HDF5 application (we want to keep a test here to check the integration).

File: applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_solver.py (hunk 14 above)
Comment (jcotela) [reply applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_solver.py:1]: I agree, but maybe not in this PR.

File: applications/FluidDynamicsApplication/python_scripts/adjoint_vmsmonolithic_solver.py (hunk 16 above)
Comment (jcotela) [reply applications/FluidDynamicsApplication/python_scripts/adjoint_vmsmonolithic_solver.py:1]: I think the long-term objective should be to skip the `AdjointFluidSolver` level so that derivation looks like `FluidSolver` -> `VMSMonolithicSolver` -> `AdjointVMSMonolithicSolver`. This would not require multiple inheritance and reduce the duplication (by skipping the extra base class, which as you say, is very similar to FluidSolver). However, we have some technical limitations (the two solvers currently have a different way of ensuring that the buffer size is properly filled).

I would propose to merge this as-is and correct the inheritance/duplication on a separate pull request, since this allows us to isolate the change in the buffer-size behavior (which has some potential for bugs).

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py (hunk 10 above)
Comment (philbucher) [reply applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py:26]: How abt sth like HDF5IOWithFluid then?

File: applications/FluidDynamicsApplication/python_scripts/adjoint_vmsmonolithic_solver.py (hunk 16 above)
Comment (philbucher) [reply applications/FluidDynamicsApplication/python_scripts/adjoint_vmsmonolithic_solver.py:1]: Sounds good

I am just suggesting, not blocking

I am saying bcs we had this problem with the chimera solver

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication_mpi.py (hunk 9 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication_mpi.py:8]: done :)

**Comment by msandre** [comment]
Re-approving. This PR is urgent for testing other PRs affecting the fluid application. I agree with @jcotela to address some of the remaining issues in a separate PR.

**Review by msandre — APPROVED** [review]
